src/WRings.sol

```diff
@@ -0,0 +1,268 @@
+    /// @param owner address of the user that owns the share tokens
+    /// @param assets amount of tokens that are being withdrawn
+    /// @param shares amount of shares that are being burned
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+    IERC20 public immutable iRings;
+    uint256 public constant fee = 1e16;
+    bytes32 public constant CALLBACK_SUCCESS = keccak256("ERC3156FlashBorrower.onFlashLoan");
+   
+
+    /// @param iRings_ ERC20 Interface for Rings token
+    constructor(IERC20 iRings_) ERC20("WRings", "WRNG", 18) {
+        iRings = iRings_;
+    }
+
+    /// @return assetTokenAddress Contract address of token utilized by the vault
+    function asset() external view returns(address){
+        return address(iRings);
+    }
+
+    /// @return totalManagedAssets Balance of the asset token within the vault
+    function totalAssets() external view returns(uint256) {
+        return _totalAssets();
+    }
+
+    /**
+     * @dev The amount of currency available to be lent.
+     * @param token The loan currency.
+     * @return The amount of `token` that can be borrowed.
+     */
+    function maxFlashLoan(address token) external view returns(uint256) {
+        require(token == address(iRings), "Token not available");
+        return _totalAssets();
+    }
+
+    /**
+     * @dev The fee to be charged for a given loan.
+     * @param token The loan currency.
+     * @param amount The amount of tokens lent.
+     * @return The amount of `token` to be charged for the loan, on top of the returned principal.
+     */
+    function flashFee(address token, uint256 amount) external view returns(uint256) {
+        require(token == address(iRings), "WRings: Unsupported token");
+        return _flashFee(amount);
+    }
+    
+    /// @notice the amount of shares the Vault would exchange for the amount of assets provided under ideal conditions
+    /// @param assets the amount of assets to send to the vault
+    /// @return shares the amount of shares received under ideal conditions
+    function convertToShares(uint256 assets) external view returns(uint256){
+        return _convertToShares(assets);
+    }
+
+    /// @notice the amount of assets the Vault would exchange for stated shares under ideal conditions
+    /// @param shares the amount of shares being exchanged
+    /// @return assets the amount of assets received under ideal conditions
+    function convertToAssets(uint256 shares) external view returns(uint256){
+        return _convertToAssets(shares);
+    }
+
+    /// @notice shows the max amount of assets that can be deposited for or by a user
+    /// @param receiver address of user that would receive the tokens
+    /// @return maxAssets max amount of assets that can be deposited
+    function maxDeposit(address receiver) external pure returns(uint256) {
+        if(receiver == address(0x0)){
```

Comment:
Why can't we deposit with `address(0)` as a `receiver`?

Reply:
To prevent tokens being burned on deposit.
This is now reflected in the deposit functions